### DISCUSSION TITLE
override pipeline settings for specific packages or filters

### DISCUSSION THREAD

**Discussion by theoephraim** [vercel/turborepo#2953] [Ideas]
Im trying to use turborepo in a monorepo that has packages built using several different tools/languages.

I'm running into issues because the tooling may be very different and need fairly different caching/dependency behaviour.

I can sort of work around it by naming npm tasks differently, but it quickly gets confusing, and it would be way nicer to have a somewhat standard set of tasks that I can run in each package in my monorepo.

This leads me to think it would be great if we could specify the pipeline config more granularly when necessary.

It seems it is possible to override settings for a specific task on a specific package (although not obvious in the docs!), but this would get very verbose and hard to maintain quickly in a larger project.

A more powerful approach would be to allow more granular settings using **filters** for the same task. Here's an example to help illustrate the idea:

```jsonc
{
  "pipeline": {
    "build": {
      // settings at the root are treated as defaults
      "dependsOn": ["^build"],
      "outputs": ["dist"],

      // optional overrides - can have multiple
      // if a package matches more than one, can just use the first match
      "overrides": [
        {
          // use filters to select package(s) to override settings
          "filters": ["foo-package"],
          // all default settings are ignored if filter matches and these are used instead
          // all existing (and future) settings can be specified here, same as at the root
          "dependsOn": ["special-foo-build-task"],
          "cache": false,
        },
        {
          // can use multiple filters, and all filter syntax options
          // maybe the git-related ones don't work... but even that maybe could be useful?
          "filters": ["./some-dir", "*glob*", "!foo"],
          //...
        },
      ]
    }
  }
}
```

This kind of setup combined with some tagging type filters (I mentioned something similar [here](https://github.com/vercel/turbo/discussions/105)) would be very powerful and flexible. Probably something like `--filter=#frontend` combined with a tags field in package json, something like `"turboTags": ["frontend", "typescript", "whatever"]` (or `"turbo": { "tags": ...`)

I haven't thought through it all too deeply, but a feature like this seems fairly straightforward to build and maintain, and I think it could help make turborepo way more flexible.

Happy to help spec it out further and implement!

**Comment by theoephraim** [comment]
looking to expand on https://github.com/vercel/turbo/discussions/105 - curious for your thoughts @jaredpalmer
Thanks!

**Comment by anthonyshew** [comment]
This can be accomplished today using [Workspace Configurations](https://turbo.build/repo/docs/core-concepts/monorepos/configuring-workspaces).
